packages/transactions/src/codecs/transaction-codec.ts

```diff
@@ -10,3 +29,80 @@ export function getNewTransactionEncoder(): VariableSizeEncoder<NewTransaction>
         ['messageBytes', getBytesEncoder()],
     ]);
 }
+
+export function getNewTransactionDecoder(): VariableSizeDecoder<NewTransaction> {
+    return transformDecoder(
+        getStructDecoder([
+            ['signatures', getArrayDecoder(fixDecoderSize(getBytesDecoder(), 64), { size: getShortU16Decoder() })],
+            ['messageBytes', getBytesDecoder()],
+        ]),
+        decodePartiallyDecodedTransaction,
+    );
+}
+
+export function getNewTransactionCodec(): VariableSizeCodec<NewTransaction> {
+    return combineCodec(getNewTransactionEncoder(), getNewTransactionDecoder());
+}
+
+type PartiallyDecodedTransaction = {
+    messageBytes: ReadonlyUint8Array;
+    signatures: ReadonlyUint8Array[];
+};
+
+function decodePartiallyDecodedTransaction(transaction: PartiallyDecodedTransaction): NewTransaction {
+    const { messageBytes, signatures } = transaction;
+
+    /*
+    Relevant message structure is at the start:
+    - transaction version
+    - transaction header (includes number of signer accounts)
+    - static accounts (signers first)
+    */
+
+    let offset = 0;
+
+    // read version
+    offset = getTransactionVersionDecoder().read(messageBytes, offset)[1];
+
+    // read header
+    // first byte of header is numSignerAccounts
+    const numSignersResult = getU8Decoder().read(messageBytes, offset);
+    const numSigners = numSignersResult[0];
+    // there are two more bytes which we don't need to read
+    offset = numSignersResult[1] + 2;
+
+    // read signer addresses (these are first in the static accounts)
+    // compact-u16 array format, read the short-u16 length
+    const addressesLength = getShortU16Decoder().read(messageBytes, offset);
+    offset = addressesLength[1];
+
+    // by decoding this array with a fixed size,
+    // we don't try to read a length prefix and only read as many addresses as we need
+    const [signerAddresses] = getArrayDecoder(getAddressDecoder(), { size: numSigners }).read(messageBytes, offset);
```

Comment:
I wonder if it wouldn't be simpler to construct a temporary decoder for this instead of manually moving the offsets.

Something like:

```ts
const signerAddressesDecoder = transformDecoder(
  getTupleDecoder([
    getTransactionVersionDecoder(),
    padRightDecoder(getU8Decoder(), 2),
    getArrayDecoder(getAddressDecoder(), { size: getShortU16Decoder() })
  ]),
  tuple => tuple[2].slice(0, tuple[1])
);
const signerAddresses = signerAddressesDecoder.decode(messageBytes);
```

I know this decodes more than it needs to (i.e. all the addresses) but maybe that's fine considering it's much easier to maintain?

Otherwise, another approach that stops decoding after signer addresses could be:

```ts
// Create a decoder that reads everything up to the beginning of the first
// signer address and returns the number of signers from the header.
const numberOfSignersDecoder = transformDecoder(
  getTupleDecoder([
    getTransactionVersionDecoder(),
    padRightDecoder(getU8Decoder(), 2),
    getShortU16Decoder(),
  ]),
  tuple => tuple[1]
);
const [numberOfSigners, signerOffset] = numberOfSignersDecoder.read(messageBytes, 0);

// Use that information to only decode the signer addresses.
const [signerAddresses] = getArrayDecoder(
  getAddressDecoder(),
  { size: numberOfSigners }
).read(messageBytes, signerOffset);
```

Reply:
This is definitely a lot cleaner! I do prefer only decoding the signer addresses because it will often be a lot less than the number of static addresses, but it probably is a bit of a premature optimisation. I do think the comments (especially with Joe's improvements) are valuable for interpreting this, but I can add them to this structure too.

Reply:
Thanks! I went with this:

https://github.com/solana-labs/solana-web3.js/blob/4214293ce77332dff899468c01962d72a306fd34/packages/transactions/src/codecs/transaction-codec.ts#L66-L77

Reply:
Perfect 👌

Reply:
Is there a reason why we wouldn't use the `getMessageHeaderDecoder` here? It seems like it's much more digestible as to what's going on. Skipping vs deserializing two bytes is not a big deal.

Reply:
The only downside of that would be that we'd have to extract `numRequiredSignatures` from the decoded header, so it'd be an extra layer there. I don't have a strong view either way though TBH.

Reply:
I'm going to merge for now, we have tests for this decoder so happy to tweak it in future!